src/shiv/bootstrap/__init__.py

```diff
@@ -208,6 +209,22 @@ def bootstrap():  # pragma: no cover
     if env.extend_pythonpath:
         extend_python_path(os.environ, sys.path.copy())
 
+    # if a preamble script was provided, run it
+    if env.preamble:
+
+        # path to the preamble
+        preamble_bin = site_packages / "bin" / env.preamble
+
+        if preamble_bin.suffix == ".py":
+            runpy.run_path(
+                preamble_bin,
+                init_globals={**globals(), **locals()},
```

Comment:
Maybe sufficient here to just send in `**locals()`?  If you send in `**globals()` do that create any confusion/problems with this script's globals and the called script's globals?

Reply:
@rsyring I was wondering about this - I wanted to include globals so that preamble authors would have access to the functions in the `bootstrap/__init__.py` module (like `cache_path`, etc). But perhaps for starters we should be more conservative and just plug in some specific stuff like `{"site_packages": site_packages, "environment": env}`. What do you think? I'm totally open to ideas here.

Reply:
I imagine you're aware of the comparable PEX functionality... In that case, the preamble _has_ to be Python, and the bootstrap script includes the code before anything else: https://github.com/pantsbuild/pex/blob/b51b340bea8a585a246997071a815d008986d6e3/pex/pex_builder.py#L447-L450

For our specific needs, we want to modify the entry point at runtime if the script is invoked with args. With PEX, we update `os.environ`. As this is currently implemented, the environment is already set up by the time the preamble script runs, so we'd instead need to modify `env` directly.

So, whichever solution you land on, we hope it supports modifying the environment so that we can dynamically set the entry point at runtime 🤞

Reply:
For reference, here's a stripped down version of our PEX preamble:
```
if os.path.islink(sys.argv[0]) and 'PEX_SCRIPT' not in os.environ:
    os.environ['PEX_SCRIPT'] = os.path.basename(sys.argv[0])
```

Reply:
> I wanted to include globals so that preamble authors would have access to the functions in the bootstrap/__init__.py module 

Shiv is included in the archive, so if someone wants access to the functions, it would make more sense IMO to just import them, e.g. `from shiv.bootstrap import cache_path`.

> But perhaps for starters we should be more conservative and just plug in some specific stuff like {"site_packages": site_packages, "environment": env}. What do you think? I'm totally open to ideas here.

You are going to be creating an API "contract" based on what variables/data are made available to the preamble script.  Easy to add things, but taking things away means a potential breaking change.  Therefore, IMO, I'd start lean and add things in the future if really needed.  Additionally, I wouldn't want to tie the name of the variables that are passed to the preamble script directly to the variable names being used in the bootstrap module and functions.   Therefore, I'd probably start with something simple like:

```python
init_globals={
    'shiv_environment': env, 
    'shiv_site_packages_path': site_packages, 
    'shiv_cache_path': site_packages.parent
})
```
I prefixed with "shiv_" since they are going to end up in the calling script's global space, less likely to have collisions that way.  I might consider UPPERCASE for the variables too.  That is usually reserved for constants, which isn't exactly true here, but I like the fact that the uppercase gives some indication that they are "special" since to the calling script, they are a bit magical...i.e. they are kind of coming out of thin air.

I would not add any of the global variables in.  IMO it's not needed and may add to confusion later.

> So, whichever solution you land on, we hope it supports modifying the environment so that we can dynamically set the entry point at runtime crossed_fingers

FWIW, this wouldn't be my preferred approach.  I want Shiv to do the work of setting up the shiv environment and extracting the packages before the preamble is called.  

For @luislew's use case, having a more flexible entry point that can "do the right thing" based on whether the script has arguments or not would be the better solution, instead of being able to modify which entry point was called with the preamble script.  

It looks like `run_path()` supports returning the global's of the called script and you could therefore permit the preamble to modify the environment here in the bootstrap function by doing something with it's return value.  I guess it just depends on whether or not you want to support that kind of use case.

Reply:
@luislew the way this PR is currently implemented would accomplish exactly what you are looking for:

```
$ cat preamble.py
import random
print("flipping a coin")
flip = random.choice([0, 1])
if flip < 1:
    os.environ['SHIV_ENTRY_POINT'] = "hello.main"
else:
    os.environ['SHIV_ENTRY_POINT'] = "hello.main2"

$ cat hello/__init__.py
def main():
    print("got 0")

def main2():
    print("got 1")

$ shiv --preamble preamble.py -o test.pyz .
Processing /private/var/folders/m1/7f1yt2xd68nf2cwr8l6lqq6c0000gn/T/tmp.pvGG45xJ
Building wheels for collected packages: hello
  Building wheel for hello (setup.py): started
  Building wheel for hello (setup.py): finished with status 'done'
  Stored in directory: /private/var/folders/m1/7f1yt2xd68nf2cwr8l6lqq6c0000gn/T/pip-ephem-wheel-cache-wbi69o_9/wheels/7a/f8/51/55dc740b08040c7d13b6aab76994923a5af676eb931cd36509
Successfully built hello
Installing collected packages: hello
Successfully installed hello-0.0.0

$ ./test.pyz
flipping a coin
got 0

$ ./test.pyz
flipping a coin
got 1
```

This is because the `env` object's properties first check in os.environ, to allow them to be overridden, so if your preamble script sets the environment variables, that would work. The only one you wouldn't be able to override in preamble is SHIV_ROOT because of the 🐔  and 🥚  problem

Reply:
@lorencarvalho One tricky problem left 😬 [`runpy.run_path`](https://docs.python.org/3/library/runpy.html#runpy.run_path) modifies `sys.argv[0]` **in place**. As a result, there's no reliable way to get a handle to the running `.pyz` file from within the preamble code. If you're running in bash you can reference `$_` but that's not a portable solution. 😞

Possible fixes:
* Expose the .pyz filepath via `env`
* Set the .pyz file in `os.environ` (e.g. `os.environ["CURRENT_SHIV"] = ...`) before invoking the preamble code
* Add the .pyz filepath to `init_globals` dict passes to `run_path`
* ...?

Any thoughts?

Reply:
@luislew 

> One tricky problem left 😬 runpy.run_path modifies sys.argv[0] in place.

Well I'll be damned, it sure does. I'm most amenable to your third suggestion of adding the .pyz's path to the init_globals that is passed to run_path. How about `archive: sys.argv[0]` ?

Reply:
I spun up the PR [here](https://github.com/linkedin/shiv/pull/159); I went with `script_name` but `archive` makes sense too.